### PR TITLE
55 allow user to withdraw their stake in either token or itoken

### PR DESCRIPTION
**Pull request by aliberkhsl**
#55

### CODE REVIEW

File: test/staking/SimpleDAIStaking.test.ts [hunk 1]
```diff
@@ -1180,6 +1188,55 @@ describe("SimpleDAISTAking - staking with cDAI mocks", () => {
...
+    const stakingAmount = ethers.utils.parseUnits("100",8);
+
+    await cDAI["mint(address,uint256)"](staker.address, stakingAmount);
+    await cDAI.connect(staker).approve(goodCompoundStaking.address, stakingAmount);
+    const productivityBeforeStake = await goodCompoundStaking.getProductivity(staker.address)
+
+    await goodCompoundStaking
+      .connect(staker)
+      .stake(stakingAmount, 100, true)
+      await goodCompoundStaking
```
Comment (sirpy) [inline test/staking/SimpleDAIStaking.test.ts:1201]: you are not testing state after stake
separate into two tests stake/withdraw

File: test/staking/SimpleDAIStaking.test.ts [hunk 2]
```diff
@@ -1180,6 +1188,55 @@ describe("SimpleDAISTAking - staking with cDAI mocks", () => {
...
+    await cDAI.connect(staker).approve(goodCompoundStaking.address, stakingAmount);
+    const productivityBeforeStake = await goodCompoundStaking.getProductivity(staker.address)
+    await goodCompoundStaking
+      .connect(staker)
+      .stake(stakingAmount, 100, true)
+      await goodCompoundStaking
+      .connect(staker)
+      .withdrawStake("1603010101010101010101", false) // 1603010101010101010101 is equaliavent of 100cDAI in DAI with currentexchange rate
+    const productivityAfterWithdraw = await goodCompoundStaking.getProductivity(staker.address)
+    expect(productivityBeforeStake[0]).to.be.equal(productivityAfterWithdraw[0])
```
Comment (sirpy) [inline test/staking/SimpleDAIStaking.test.ts:1221]: again you just test final state, but maybe nothing happened in between at all

File: test/staking/SimpleDAIStaking.test.ts [hunk 3]
```diff
@@ -1180,6 +1188,55 @@ describe("SimpleDAISTAking - staking with cDAI mocks", () => {
...
+    await goodCompoundStaking
+      .connect(staker)
+      .stake(stakingAmount, 100, false)
+    await goodCompoundStaking
+      .connect(staker)
+      .withdrawStake("623826387", true) 
+    await goodCompoundStaking.connect(staker).withdrawStake("000000000036236363637",false);// 000000000036236363637 is precision loss due to itoken decimals < token decimals
+    const productivityAfterWithdraw = await goodCompoundStaking.getProductivity(staker.address)
+    
+    expect(productivityBeforeStake[0]).to.be.equal(productivityAfterWithdraw[0])
```
Comment (sirpy) [inline test/staking/SimpleDAIStaking.test.ts:1238]: test should always check the action expected result.
you stake, you expect user balance to be less and you expect contract balance+records to have changed

**Review by sirpy — CHANGES_REQUESTED** [review]

File: contracts/staking/SimpleStaking.sol [hunk 4]
```diff
@@ -77,61 +77,97 @@ contract SimpleStaking is AbstractGoodStaking, StakingToken {
...
+				"transferFrom failed, make sure you approved token transfer"
+			);
+			_amount = iTokenWorthinToken(_amount);
+		} else {
+			require(
+				token.transferFrom(msg.sender, address(this), _amount),
+				"transferFrom failed, make sure you approved token transfer"
+			);
+			// approve the transfer to defi protocol
+			token.approve(address(iToken), _amount);
```
[the remark this answers is not in the document]
Comment (aliberkhsl) [reply contracts/staking/SimpleStaking.sol:116]: So in the constructor maybe?

File: test/staking/SimpleDAIStaking.test.ts (hunk 2 above)
Comment (aliberkhsl) [reply test/staking/SimpleDAIStaking.test.ts:1243]: I added proper variables that test it also

File: test/staking/SimpleDAIStaking.test.ts (hunk 3 above)
Comment (aliberkhsl) [reply test/staking/SimpleDAIStaking.test.ts]: I added proper variables that test it also

File: contracts/staking/SimpleStaking.sol [hunk 5]
```diff
@@ -77,61 +77,97 @@ contract SimpleStaking is AbstractGoodStaking, StakingToken {
...
+			_donationPer == 0 || _donationPer == 100,
+			"Donation percentage should be 0 or 100"
 		);
+		require(_amount > 0, "You need to stake a positive token amount");
+		if (_inInterestToken) {
+			require(
+				iToken.transferFrom(msg.sender, address(this), _amount),
+				"transferFrom failed, make sure you approved token transfer"
+			);
+			_amount = iTokenWorthinToken(_amount);
```
[the remark this answers is not in the document]
Comment (aliberkhsl) [reply contracts/staking/SimpleStaking.sol:108]: replaced it

File: contracts/staking/SimpleStaking.sol [hunk 6]
```diff
@@ -77,61 +77,97 @@ contract SimpleStaking is AbstractGoodStaking, StakingToken {
...
 		require(
-			token.transferFrom(msg.sender, address(this), _amount),
-			"transferFrom failed, make sure you approved token transfer"
+			_donationPer == 0 || _donationPer == 100,
+			"Donation percentage should be 0 or 100"
 		);
+		require(_amount > 0, "You need to stake a positive token amount");
+		if (_inInterestToken) {
+			require(
+				iToken.transferFrom(msg.sender, address(this), _amount),
```
[the remark this answers is not in the document]
Comment (aliberkhsl) [reply contracts/staking/SimpleStaking.sol]: changed it

File: contracts/staking/SimpleStaking.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (sirpy) [reply contracts/staking/SimpleStaking.sol]: yes

**Comment by sirpy** [comment]
what else is required here?

File: contracts/staking/SimpleStaking.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (aliberkhsl) [reply contracts/staking/SimpleStaking.sol]: Already did it token.approve(address(iToken), type(uint256).max);

**Pull request merged**